File: apps/android/src/main/kotlin/space/bitos/app/ui/feed/ImportMediaSheet.kt
package space.bitos.app.ui.feed

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.selection.selectable
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Button
import androidx.compose.material3.LinearProgressIndicator
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.RadioButton
import androidx.compose.material3.Surface
import androidx.compose.material3.Switch
import androidx.compose.material3.Text
import androidx.compose.ui.draw.clip
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import space.bitos.app.ui.theme.BitOSColors
import space.bitos.app.ui.theme.BitOSSpacing

/** Picked media awaiting publish. */
data class PickedMedia(
    val bytes: ByteArray,
    val mimeType: String,
)

enum class MediaPublishPhase { PICK, UPLOADING, PUBLISHING, DONE }

data class MediaPublishUiState(
    val picked: PickedMedia? = null,
    val phase: MediaPublishPhase = MediaPublishPhase.PICK,
    val failure: String? = null,
)

/** MST-017 meme lane: render → upload → kind-20 phases (no pick phase). */
enum class MemePublishPhase { IDLE, UPLOADING, PUBLISHING, DONE }

data class MemePublishUiState(
    val phase: MemePublishPhase = MemePublishPhase.IDLE,
    val failure: String? = null,
)

/**
 * Media import → publish sheet (CAP-005 + PUB media path): gallery pick →
 * caption → hash-verified Blossom upload → kind-22 through the receipt
 * machine. Failures (oversized files, signer refusal, upload mismatch) are
 * surfaced, never swallowed.
 */
@Composable
fun ImportMediaContent(
    state: MediaPublishUiState,
    onPublish: (caption: String, altText: String, contentWarningReason: String?) -> Unit,
    onPick: (android.net.Uri, android.content.ContentResolver) -> Unit,
    onCancel: () -> Unit,
) {
    val context = LocalContext.current
    var caption by remember { mutableStateOf("") }
    var altText by remember { mutableStateOf("") }
    var contentWarningOn by remember { mutableStateOf(false) }
    var contentWarningReason by remember { mutableStateOf(ContentWarningReasons.DEFAULT) }
    val picker = androidx.activity.compose.rememberLauncherForActivityResult(
        contract = androidx.activity.result.contract.ActivityResultContracts.GetContent(),
    ) { uri -> uri?.let { onPick(it, context.contentResolver) } }

    Column(
        modifier = Modifier
            .fillMaxWidth()
            .padding(horizontal = BitOSSpacing.screen)
            .padding(bottom = BitOSSpacing.xl),
        verticalArrangement = Arrangement.spacedBy(BitOSSpacing.md),
    ) {
        Row(verticalAlignment = Alignment.CenterVertically) {
            Text("New video", style = MaterialTheme.typography.headlineMedium)
            Spacer(Modifier.weight(1f))
            space.bitos.app.ui.components.SheetCloseIcon(onClose = onCancel)
        }

        state.failure?.let { failure ->
            Surface(shape = RoundedCornerShape(12.dp), color = BitOSColors.surface) {
                Text(
                    failure,
                    style = MaterialTheme.typography.bodySmall,
                    color = BitOSColors.error,
                    modifier = Modifier.padding(BitOSSpacing.base),
                )
            }
        }

        when (state.phase) {
            MediaPublishPhase.PICK -> {
                if (state.picked == null) {
                    Button(onClick = { picker.launch("video/*") }, modifier = Modifier.fillMaxWidth()) {
                        Text("Pick a video from your library")
                    }
                } else {
                    PickedSummary(state.picked!!)
                    // Post details (reference scr-details): counter + alt +
                    // content warning ride the existing kind-22 pipeline.
                    OutlinedTextField(
                        value = caption,
                        onValueChange = { if (it.length <= 2000) caption = it },
                        placeholder = { Text("Add a caption…") },
                        supportingText = { Text("${caption.length} / 2000") },
                        minLines = 2,
                        modifier = Modifier.fillMaxWidth(),
                    )
                    OutlinedTextField(
                        value = altText,
                        onValueChange = { if (it.length <= 200) altText = it },
                        placeholder = { Text("Describe the video for screen readers…") },
                        supportingText = {
                            Text("Alt text · required by your policy (${altText.length}/200)")
                        },
                        minLines = 2,
                        modifier = Modifier.fillMaxWidth(),
                    )
                    ContentWarningSection(
                        enabled = contentWarningOn,
                        reason = contentWarningReason,
                        onToggle = { contentWarningOn = it },
                        onReason = { contentWarningReason = it },
                    )
                    Row(horizontalArrangement = Arrangement.spacedBy(BitOSSpacing.sm)) {
                        OutlinedButton(onClick = { picker.launch("video/*") }) { Text("Change") }
                        Button(
                            onClick = {
                                onPublish(caption, altText, if (contentWarningOn) contentWarningReason else null)
                            },
                            enabled = caption.isNotBlank() || altText.isNotBlank(),
                        ) {
                            Text("Upload & publish")
                        }
                    }
                }
            }
            MediaPublishPhase.UPLOADING -> {
                Text("Uploading (hash-verified)…", style = MaterialTheme.typography.titleMedium, fontWeight = FontWeight.W600)
                LinearProgressIndicator(color = BitOSColors.primary, modifier = Modifier.fillMaxWidth())
                state.picked?.let { PickedSummary(it) }
            }
            MediaPublishPhase.PUBLISHING -> {
                Text("Publishing to relays…", style = MaterialTheme.typography.titleMedium, fontWeight = FontWeight.W600)
                LinearProgressIndicator(color = BitOSColors.primary, modifier = Modifier.fillMaxWidth())
            }
            MediaPublishPhase.DONE -> {
                Text(
                    "Published ✓ — your video appears in the feed once relays confirm.",
                    style = MaterialTheme.typography.titleMedium,
                    color = BitOSColors.success,
                    fontWeight = FontWeight.W600,
                )
                OutlinedButton(onClick = onCancel) { Text("Done") }
            }
        }
    }
}

@Composable
private fun PickedSummary(picked: PickedMedia) {
    Surface(shape = RoundedCornerShape(12.dp), color = BitOSColors.surface) {
        Column(Modifier.padding(BitOSSpacing.base)) {
            Text("${picked.mimeType} • ${picked.bytes.size / (1024 * 1024)}MB", style = MaterialTheme.typography.bodySmall)
        }
    }
}

/** NIP-36 reason choices (reference scr-details select). */
internal object ContentWarningReasons {
    const val DEFAULT = "Flashing imagery"
    val ALL = listOf("Flashing imagery", "Sensitive topic", "Loud audio")
}

/** Content warning gate (reference scr-details): toggle + reason picker. */
@Composable
private fun ContentWarningSection(
    enabled: Boolean,
    reason: String,
    onToggle: (Boolean) -> Unit,
    onReason: (String) -> Unit,
) {
    Surface(shape = RoundedCornerShape(12.dp), color = BitOSColors.surface) {
        Column(Modifier.padding(BitOSSpacing.base), verticalArrangement = Arrangement.spacedBy(BitOSSpacing.sm)) {
            Row(verticalAlignment = Alignment.CenterVertically) {
                Column(Modifier.weight(1f)) {
                    Text("Content warning", style = MaterialTheme.typography.titleSmall)
                    Text(
                        "Gate playback behind a visible warning",
                        style = MaterialTheme.typography.bodySmall,
                        color = BitOSColors.textSecondary,
                    )
                }
                Switch(checked = enabled, onCheckedChange = onToggle)
            }
            if (enabled) {
                ContentWarningReasons.ALL.forEach { option ->
                    Row(
                        verticalAlignment = Alignment.CenterVertically,
                        modifier = Modifier
                            .fillMaxWidth()
                            .clip(RoundedCornerShape(8.dp))
                            .selectable(selected = option == reason) { onReason(option) }
                            .padding(vertical = BitOSSpacing.xs),
                    ) {
                        RadioButton(selected = option == reason, onClick = { onReason(option) })
                        Text(option, style = MaterialTheme.typography.bodyMedium)
                    }
                }
            }
        }
    }
}

File: apps/ios/BitOS/Features/Home/ImportMediaSheet.swift
import BusinessCore
import PhotosUI
import SwiftUI

enum MediaPublishPhase: Sendable, Equatable { case pick, uploading, publishing, done }

struct MediaPublishUiState: Sendable, Equatable {
    var phase: MediaPublishPhase = .pick
    var failure: String?
    var busy = false
}

/**
 * Media import → publish sheet (CAP-005 + PUB media path): photo-library
 * pick → caption → hash-verified Blossom upload → kind-22 through the
 * receipt machine. Oversized files, signer refusal and hash mismatches are
 * surfaced, never swallowed.
 */
struct ImportMediaSheet: View {
    let onClose: () -> Void
    /** Freshly captured take from the camera flow (record → trim); the
     *  sheet opens straight into the caption/publish state for it. */
    var capturedData: Data? = nil
    var capturedMime: String = "video/mp4"
    @Environment(AppEnvironment.self) private var environment
    @Environment(IdentityStore.self) private var identity
    @State private var state = MediaPublishUiState()
    @State private var caption = ""
    @State private var altText = ""
    @State private var contentWarningOn = false
    @State private var contentWarningReason = Self.contentWarningReasons.first ?? "Flashing imagery"
    @State private var pickerItem: PhotosPickerItem?
    @State private var pickedData: Data?
    @State private var pickedMime = "video/mp4"
    @State private var seededCapture = false

    /// NIP-36 reason choices (reference scr-details select).
    private static let contentWarningReasons = ["Flashing imagery", "Sensitive topic", "Loud audio"]

    private let maxBytes = 64 * 1024 * 1024

    var body: some View {
        NavigationStack {
            content
                .padding(BitOSTheme.Spacing.screen)
                .background(BitOSTheme.background)
                .navigationTitle("New video")
                .navigationBarTitleDisplayMode(.inline)
                .toolbar {
                    ToolbarItem(placement: .cancellationAction) { SheetCloseButton(action: onClose) }
                }
                .onAppear {
                    guard !seededCapture, let capturedData else { return }
                    seededCapture = true
                    if capturedData.count > maxBytes {
                        state.failure = "Recording exceeds the 64MB limit."
                    } else {
                        pickedData = capturedData
                        pickedMime = capturedMime
                    }
                }
        }
        .preferredColorScheme(BitOSTheme.preferredScheme)
    }

    @ViewBuilder
    private var content: some View {
        VStack(spacing: BitOSTheme.Spacing.md) {
            if let failure = state.failure {
                Text(failure).font(.footnote).foregroundStyle(BitOSTheme.error)
            }
            switch state.phase {
            case .pick:
                let pickLabel = pickedData == nil ? "Pick a video from your library" : "Change video"
                PhotosPicker(selection: $pickerItem, matching: .videos) {
                    Label(pickLabel, systemImage: "photo.on.rectangle")
                }
                .buttonStyle(.borderedProminent)
                .tint(BitOSTheme.accent)
                .onChange(of: pickerItem) { _, item in
                    guard let item else { return }
                    Task { @MainActor in
                        if let data = try? await item.loadTransferable(type: Data.self) {
                            if data.count > maxBytes {
                                state.failure = "Media exceeds the 64MB limit."
                                pickedData = nil
                            } else {
                                pickedData = data
                                state.failure = nil
                            }
                        }
                    }
                }
                if let data = pickedData {
                    Text("\(pickedMime) • \(data.count / 1024 / 1024)MB")
                        .font(.caption)
                        .foregroundStyle(BitOSTheme.textSecondary)
                    // Post details (reference scr-details): counter + alt +
                    // content warning ride the existing kind-22 pipeline.
                    VStack(alignment: .leading, spacing: 2) {
                        TextField("Add a caption…", text: $caption)
                            .textFieldStyle(.roundedBorder)
                        Text("\(caption.count) / 2000")
                            .font(.caption2)
                            .foregroundStyle(BitOSTheme.textTertiary)
                    }
                    VStack(alignment: .leading, spacing: 2) {
                        TextEditor(text: $altText)
                            .font(.body)
                            .frame(minHeight: 56)
                            .scrollContentBackground(.hidden)
                            .background(BitOSTheme.surface, in: RoundedRectangle(cornerRadius: 8))
                            .accessibilityLabel("Alt text")
                        Text("Alt text · required by your policy (\(altText.count)/200)")
                            .font(.caption2)
                            .foregroundStyle(BitOSTheme.textTertiary)
                    }
                    ContentWarningSection(
                        on: $contentWarningOn,
                        reason: $contentWarningReason,
                        options: Self.contentWarningReasons
                    )
                    Button("Upload & publish") {
                        Task { await publish() }
                    }
                    .buttonStyle(.borderedProminent)
                    .tint(BitOSTheme.accent)
                    .disabled(
                        (caption.trimmingCharacters(in: .whitespaces).isEmpty &&
                            altText.trimmingCharacters(in: .whitespaces).isEmpty) || state.busy
                    )
                }
            case .uploading:
                ProgressView().tint(BitOSTheme.accent)
                Text("Uploading (hash-verified)…").font(.subheadline)
            case .publishing:
                ProgressView().tint(BitOSTheme.accent)
                Text("Publishing to relays…").font(.subheadline)
            case .done:
                Label("Published ✓", systemImage: "checkmark.circle.fill")
                    .foregroundStyle(BitOSTheme.success)
                Button("Done") { onClose() }
                    .buttonStyle(.borderedProminent)
                    .tint(BitOSTheme.accent)
            }
            Spacer()
        }
    }

    private func publish() async {
        guard let data = pickedData else { return }
        state.busy = true
        state.phase = .uploading
        state.failure = nil
        do {
            let bridge = (environment.businessCore as? FrameworkBusinessCoreClient)?.bridgeForFollowing() ?? BusinessCoreBridge()
            let uploaded = try await BlossomUploader(bridge: bridge).upload(
                bytes: data, mimeType: pickedMime, identity: identity,
                serverUrl: "https://blossom.primal.net"
            )
            state.phase = .publishing
            await environment.notePublisher.publishMediaNote(
                caption: caption,
                altText: altText,
                contentWarningReason: contentWarningOn ? contentWarningReason : nil,
                url: uploaded.url, hash: uploaded.hash, mime: uploaded.mime, size: uploaded.size
            )
            state.phase = .done
        } catch {
            state.phase = .pick
            state.failure = error.localizedDescription
        }
        state.busy = false
    }
}

/// Content warning gate (reference scr-details): toggle + reason picker.
private struct ContentWarningSection: View {
    @Binding var on: Bool
    @Binding var reason: String
    let options: [String]

    var body: some View {
        VStack(alignment: .leading, spacing: BitOSTheme.Spacing.sm) {
            HStack {
                VStack(alignment: .leading, spacing: 2) {
                    Text("Content warning").font(.subheadline.weight(.semibold))
                    Text("Gate playback behind a visible warning")
                        .font(.caption)
                        .foregroundStyle(BitOSTheme.textSecondary)
                }
                Spacer()
                Toggle("", isOn: $on)
                    .labelsHidden()
                    .accessibilityLabel("Content warning")
            }
            if on {
                Picker("Reason", selection: $reason) {
                    ForEach(options, id: \.self) { option in
                        Text(option).tag(option)
                    }
                }
                .pickerStyle(.segmented)
            }
        }
        .padding(BitOSTheme.Spacing.card)
        .background(BitOSTheme.surface, in: RoundedRectangle(cornerRadius: BitOSTheme.Radius.md))
    }
}
